ownCloud/Client/Sharing/GroupSharingEditTableViewController.swift

```diff
@@ -0,0 +1,308 @@
+		section.add(row: StaticTableViewRow(toggleItemWithAction: { [weak self] (row, _) in
+			if let self = self, let selected = row.value as? Bool {
+				self.changePermissions(enabled: selected, permissions: [ permission ], completionHandler: {(_) in
+					self.hidePermissionsIfNeeded()
+				})
+			}
+		}, title: title.localized, subtitle: "", selected: selected, identifier: identifier))
+	}
+
+	func addPermissionEditSection(animated : Bool = false) {
+		let section = StaticTableViewSection(headerTitle: nil, footerTitle: nil, identifier: "permission-edit-section")
+		guard let share = share else { return }
+
+		self.addPermissionRow(to: section, with: "Can Create", permission: .create, selected: share.canCreate, identifier: "permission-section-edit-create")
+		self.addPermissionRow(to: section, with: "Can Change", permission: .update, selected: share.canUpdate, identifier: "permission-section-edit-change")
+		self.addPermissionRow(to: section, with: "Can Delete", permission: .delete, selected: share.canDelete, identifier: "permission-section-edit-delete")
+
+		let subtitles = [
+			"Allows the users you share with to create new files and add them to the share".localized,
+			"Allows uploading a new version of a shared file and replacing it".localized,
+			"Allows the users you share with to delete shared files".localized
+		]
+		updateSubtitles(subtitles: subtitles, section: section)
+
+		self.insertSection(section, at: 1, animated: animated)
+	}
+
+	func hidePermissionsIfNeeded() {
+		if share?.canDelete == false, share?.canCreate == false, share?.canUpdate == false {
+			OnMainThread {
+				let section = self.sectionForIdentifier("permission-section")
+				let newRow = section?.row(withIdentifier: "permission-section-edit")
+				newRow?.cell?.accessoryType = .none
+				newRow?.value = false
+
+				if let section = self.sectionForIdentifier("permission-edit-section") {
+					self.removeSection(section, animated: true)
+				}
+			}
+		}
+	}
+
+	func changePermissions(enabled: Bool, permissions : [OCSharePermissionsMask], completionHandler: @escaping (_ error : Error?) -> Void ) {
+		guard let share = share else { return }
+
+		if createShare {
+			for permissionValue in permissions {
+				if enabled {
+					permissionMask?.insert(permissionValue)
+				} else {
+					permissionMask?.remove(permissionValue)
+				}
+			}
+		} else {
+			if let core = self.core {
+				core.update(share, afterPerformingChanges: {(share) in
+					for permissionValue in permissions {
+						if enabled {
+							share.permissions.insert(permissionValue)
+						} else {
+							share.permissions.remove(permissionValue)
+						}
+					}
+				}, completionHandler: { (error, share) in
+					if error == nil {
+						guard let changedShare = share else { return }
+						self.share?.permissions = changedShare.permissions
```

Comment:
After completing an update, the SDK fetches the shares metadata anew, creating a completely new share that exactly represents the share on the server.

IMO it'd therefore be better to use that `OCShare` instance and set it as the view controller's `share`. A `didSet` block on the view controller's `share` could then update the view controller as needed to match the updated share.

I write this only once here, but it also applies to all similar uses of `core.update(share, afterPerformingChanegs:,completionHandler:)` in this view controller and also `PublicLinkEditTableViewController`.

Reply:
IMP would should move this to later, because this needs more time to reorganize the current view.

Reply:
Agreed!